k4GaudiPandora/include/DDPandoraPFANewAlgorithm.h

```diff
@@ -0,0 +1,342 @@
+   *  @param  lhCalCollections CalorimeterHit Collection for the HLCal
+   *  @param  caloLinkCollections the associations between CalorimeterHits and MCParticles
+   * 
+   *  @return tuple of reconstructed: (Clusters, RecoParticles, Verticies)
+   *  
+   */
+  std::tuple<edm4hep::ClusterCollection, edm4hep::ReconstructedParticleCollection, edm4hep::VertexCollection> operator()(
+   const std::vector<const edm4hep::MCParticleCollection*>& MCParticleCollections,
+   const std::vector<const edm4hep::VertexCollection*>& kinkCollections,
+   const std::vector<const edm4hep::VertexCollection*>& prongCollections,
+   const std::vector<const edm4hep::VertexCollection*>& splitCollections,
+   const std::vector<const edm4hep::VertexCollection*>& v0Collections,
+   const std::vector<const edm4hep::TrackCollection*>& trackCollections,
+   const std::vector<const edm4hep::TrackerHitSimTrackerHitLinkCollection*>& trackerHitLinkCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& eCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& hCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& mCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& lCalCollections,
+   const std::vector<const edm4hep::CalorimeterHitCollection*>& lhCalCollections,
+   const std::vector<const edm4hep::CaloHitSimCaloHitLinkCollection*>& caloLinkCollections
+ ) const override;
+
+
+
+  /**
+     *  @brief  End, called at shutdown
+     */
+  StatusCode finalize();
+
+  /**
+     *  @brief  Get address of the pandora instance
+     *
+     *  @return address of the pandora instance
+     */
+  const pandora::Pandora* GetPandora() const;
+
+  /**
+     *  @brief  Get address of the current lcio event
+     *
+     *  @param  pPandora address of the relevant pandora instance
+     *
+     *  @return address of the current lcio event
+     */
+  //static const EVENT::LCEvent* GetCurrentEvent(const pandora::Pandora* const pPandora);
+
+private:
+  /**
+     *  @brief  Register user algorithm factories, energy correction functions and particle id functions,
+     *          insert user code here
+     */
+  pandora::StatusCode RegisterUserComponents() const;
+
+  /**
+     *  @brief  Copy some steering parameters between settings objects
+     */
+  void FinaliseSteeringParameters();
+
+  /**
+     *  @brief  Reset the pandora pfa new processor
+     */
+  void Reset() const;
+
+  pandora::Pandora*    m_pPandora             = NULL;  ///< Address of the pandora instance
+  DDCaloHitCreator*    m_pCaloHitCreator      = NULL;  ///< The calo hit creator
+  DDGeometryCreator*   m_pGeometryCreator     = NULL;  ///< The geometry creator
+  DDTrackCreatorBase*  m_pTrackCreator        = NULL;  ///< The track creator
+  DDMCParticleCreator* m_pDDMCParticleCreator = NULL;  ///< The mc particle creator
+  DDPfoCreator*        m_pDDPfoCreator        = NULL;  ///< The pfo creator
```

Comment:
These could probably all be `unique_ptr` (ideally), that would also mean that we could entirely remove the `finalize` method, which only cleans them up at the moment.